Improve JavaScript test support

TW80000 opened the pull request:
* JS implementation of `test` method
* JS implementation of `print` method
* Reset number of passed/failed tests on each run
* Track duration of test execution
* Track last executed time

Commit:
Add better JS test support

TW80000 commented:
@kgrgreer @shreymahendru Please review.

Commit:
Count errors thrown in test as a failure

shreymahendru commented:
@TW80000 The runScript/runTest method should be wrapped in a promise, or it should return a promise where ``eval`` is wrapped in a ``Promise.resolve``. This is because when a user is writing a script/test, it could have a Promise. Eg: a select/find on a DAO, your scripts' output should be the response to that promise and also once the promise is resolved, update the script's status and output.
And wrapping synchronous code in a promise has no effect, so it would work the same for when the user doesn't have a promise in their script.

TW80000 commented:
> This is because when a user is writing a script/test, it could have a Promise. Eg: a select/find on a DAO, your scripts' output should be the response to that promise and also once the promise is resolved, update the script's status and output.

~~It already works that way as is. I can include asynchronous code in a test that calls the `test()` method in a callback and it will update `passed` and `failed` when the test completes and show the output as well.~~

I see what you mean now. Will fix.

shreymahendru commented on src/foam/nanos/script/Script.js:
```diff
@@ -227,7 +227,7 @@ foam.CLASS({
           with ( { log: log, print: log, x: self.__context__ } ) {
             this.status = this.ScriptStatus.RUNNING;
-            var ret = eval(this.code);
+            this.runScript(x);
```
If you are calling ``this.runScript`` here you need to have a js ``runScript`` function in Script.js, which would have an ``eval``

shreymahendru commented on src/foam/nanos/test/Test.js:
```diff
@@ -88,6 +99,9 @@ foam.CLASS({
     {
       name: 'print',
+      code: function(message) {
+        this.output += message + '\n';
```
This ``print`` method should be in Script.js

shreymahendru commented on src/foam/nanos/test/Test.js:
```diff
@@ -100,6 +114,25 @@ foam.CLASS({
+        var endTime = Date.now();
+        var duration = endTime - startTime; // Unit: milliseconds
+
+        this.lastRun = new Date();
```
You need to do a ``put`` to the dao once the script has finished running.

Commit:
Fix JS tests

They will now complete when the asynchronous operations going on in them
finish.

Commit:
Remove unused function

Commit:
Put to scriptDAO on test completion

Commit:
Fix print function in JS tests

Commit:
Fix print function in JS scripts
